[BOLT][AArch64] Support cdsplit for AArch64

liusy58 opened the pull request:
CDSplit enables BB to be cold, warm or hot. So this assert should be removed.

liusy58 committed:
[BOLT] BB can be cold, warm and hot.

CDSplit enables BB to be cold, warm or hot. So this assert should be removed.

paschalis-mpeis commented:
Indeed, I remember seeing this assert in some instances when using `--split-strategy=cdsort` or `--split-strategy=all` on AArch64, but probably it also exists on other targets.

However, this would require more changes and verifying whether the prior logic is still sound. For example:
- the `setIsCold` (and it's users) is also not aware of 'warm' blocks.
- LongJmp is using it for creating/caching stubs and finding the hot/cold frontier
- sizes estimation in BinaryFunction
- a few more passes and metrics might also be incompatible

liusy58 commented:
Yeah, In fact, I have implemented a new LongJmp for `cdsplit`. I will add them later. Since it will be a big patch, so I will add step by step.

paschalis-mpeis reviewed (commented):
Thanks @liusy58. Yeap, these changes are likely steps in the right direction.

However, my understanding is that this assert stays there until all usages of `isCold` and `setIsCold` can also amount for 'warm' blocks. Perhaps if this is done step by step, the asserts can be hoisted at the callers of `isCold` that remain unhandled?

liusy58 committed:
[BOLT][AArch64] Support cdsplit for AArch64.

The current LongJmp pass only supports hot-cold splitting and is
unaware of the warm part. This oversight results in missing stubs
that need to be inserted.  Add code for warm layout.

paschalis-mpeis reviewed (commented):
Hey @liusy58,

Thanks for adding support for `cdsplit` to LongJmp. It looks like you have created local/global warm stub groups.

The assertion you [aim to eliminate](https://github.com/llvm/llvm-project/pull/121475/files#diff-89ce892ef20346fe79278dbafccad479feb00449e85d02772413a8386655f37eL675) in `isCold` is not AArch64-specific and LongJmp is not the only area affected. I believe it'll be hard to do all these changes in one go, which is why I think that more PRs will be probably needed. What Meta thinks here? (@ShatianWang, @maksfb)

For example using `-split-strategy=cdsplit --dump-dot-all` should crash regardless of architecture. On X86 one can indeed get a binary with hot/warm/cold sections, but I'm not sure to which extend this is due to luck. I would expect under some scenarios to hit this assertion with ICP, ShrinkWrapping, and TailDuplication, ..

Now on AArch64, the crash happens in most binaries of interest. That is because `LongJmp` always runs and when it adds some stubs the assertion is triggered. So we definitely want to support it.

#### Noting AArch64-specific behaviour with ASSERTIONS off:

(given how much functionality is missing, this behaviour is not surprising; noting only for reference)

If one uses llvm-bolt with ASSERTIONS Off, chances are they will get a number of 'fixup' errors being out of range and ultimately LongJmp will stuck, probably in some sort of an infinite loop.

BTW those fixup errors happen when the AsmBackend cannot adjustFixupValue and they are not crashing bolt:
```
<unknown>:0: error: fixup value out of range
```


## Next steps on this PR?
We will need Meta opinion on this. Given we agree on multiple PRs, I made a suggestion below to keep the assertion but make it conditional. LongJmp can then come as a separate PR. More PRs will be needed for the other areas.

liusy58 commented on bolt/include/bolt/Core/BinaryBasicBlock.h:
```diff
@@ -676,7 +676,7 @@ class BinaryBasicBlock {
   bool isWarm() const { return Fragment == FragmentNum::warm(); }
 
-  bool isHot() const { return Fragment == FragmentNum::main(); }
+  bool isMain() const { return Fragment == FragmentNum::main(); }
```
I think main is not equal to `hot`. For a cold function, if it is not split, all the BBs are in the main fragment, which doesn't mean hot.

paschalis-mpeis commented:
> For example using -split-strategy=cdsplit --dump-dot-all should crash regardless of architecture. On X86 one can indeed get a binary with hot/warm/cold sections, but I'm not sure to which extend this is due to luck. I would expect under some scenarios to hit this assertion with ICP, ShrinkWrapping, and TailDuplicatio

Regarding my earlier comment, I reference issue #123835, which is a crash with `ShrinkWrapping` and `cdsplit` on x86. In that particular case the reproducer is not deterministic. @aaupov [suggested](https://github.com/llvm/llvm-project/issues/123835#issuecomment-2605925949) to either disable one of the two optimizations.

I also became aware of reports where with Assertions=Off, `LongJmp` stucks due to `cdsplit` but the reasons were not obvious to end users. I've added a note on this earlier, for referencing.

IMHO, it's worth proceeding with a number of PRs here to gradually clear up the space of all `isCold` users.

paschalis-mpeis commented:
Closing as we do support cdsplit on compact mode with no plans to support it on LongJmp.

Either way, thanks for the effort @liusy58
